includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp-notes.php

```diff
@@ -0,0 +1,146 @@
+		$mc = Newspack_Newsletters_Mailchimp::instance();
+		$audience_ids = [];
+		foreach ( $lists as $list ) {
+			if ( Subscription_List::is_local_form_id( $list ) ) {
+				continue;
+			}
+			$remote_list_details = $mc->maybe_extract_group_or_tag_list( $list );
+			if ( false !== $remote_list_details ) {
+				$audience_ids[] = $remote_list_details['list_id'];
+				continue;
+			}
+			$audience_ids[] = $list;
+		}
+		return $audience_ids;
+	}
+
+	/**
+	 * Handles the upsert of a contact.
+	 *
+	 * @param string         $provider The provider name.
+	 * @param array          $contact  {
+	 *    Contact information.
+	 *
+	 *    @type string   $email    Contact email address.
+	 *    @type string   $name     Contact name. Optional.
+	 *    @type string[] $metadata Contact additional metadata. Optional.
+	 * }
+	 * @param string[]|false $lists    Array of list IDs to subscribe the contact to.
+	 * @param array|WP_Error $result   Array with data if the contact was added or error if failed.
+	 * @param bool           $is_updating Whether the contact is being updated. If false, the contact is being created.
+	 * @param string         $context  Context of the update for logging purposes.
+	 * @return void
+	 */
+	public static function handle_upsert( $provider, $contact, $lists, $result, $is_updating, $context ) {
+		if ( 'mailchimp' !== $provider ) {
+			return;
+		}
+
+		$lists_string = implode( ', ', $lists );
+		$message = sprintf(
+			/* translators: 1: email address, 2: list IDs */
+			__( 'Contact updated by Newspack from site %1$s. Context: %2$s. Lists added: %3$s', 'newspack-newsletters' ),
+			get_site_url(),
+			$context,
+			$lists_string
+		);
+
+		self::add_note( $contact['email'], $lists, $message );
+	}
+
+	/**
+	 * Handles the update of lists for a contact.
+	 *
+	 * Note that we will add a note even to the lists (audiences) that the contact is being removed from.
+	 *
+	 * @param string        $provider        The provider name.
+	 * @param string        $email           Contact email address.
+	 * @param string[]      $lists_to_add    Array of list IDs to subscribe the contact to.
+	 * @param string[]      $lists_to_remove Array of list IDs to remove the contact from.
+	 * @param bool|WP_Error $result          True if the contact was updated or error if failed.
+	 * @param string        $context         Context of the update for logging purposes.
+	 * @return void
+	 */
+	public static function handle_update_lists( $provider, $email, $lists_to_add, $lists_to_remove, $result, $context ) {
+		if ( 'mailchimp' !== $provider ) {
+			return;
+		}
+
+		$lists_to_add_string = implode( ', ', $lists_to_add );
+		$lists_to_remove_string = implode( ', ', $lists_to_remove );
+
+		$message = sprintf(
+			/* translators: 1: email address, 2: lists added, 3: lists removed */
+			__( 'Contact updated by Newspack. Context: %1$s. Lists added: %2$s. Lists removed: %3$s', 'newspack-newsletters' ),
+			$context,
+			$lists_to_add_string,
```

Comment:
```suggestion
			$lists_to_add_string ? $lists_to_add_string : '(none)',
```

Reply:
I'm not sure if `(none)` is the best when it's empty, but currently this comes through as e.g.
```
Contact updated by Newspack. Context: User updated their subscriptions on My Account page. Lists added: . Lists removed: 62688a1068
```
and
```
Contact updated by Newspack. Context: User updated their subscriptions on My Account page. Lists added: 62688a1068. Lists removed:
```

Reply:
addressed in https://github.com/Automattic/newspack-newsletters/pull/1574/commits/15ea318512af34a965810a64a9c916ebf453ecde